File: src/facemesh_mouse/modules/point_tracker.py
"""Optical-flow point tracking for cursor movement.

Tracks several points on rigid parts of the face (nose bridge, nostrils,
temples, cheek edges -- deliberately not the mouth/eyebrows/eyelids, which
move during gestures) frame-to-frame with Lucas-Kanade optical flow, and
reports the average of their movement. Averaging multiple points cancels
noise that following a single face landmark carries straight through to
the cursor: any point that has drifted or lost tracking is pruned before
the average is taken, so an outlier can't drag the whole result off
course.

Distance thresholds scale with the tracked head's size in the frame rather
than using fixed pixel counts, so the same relative behavior holds whether
the user sits close to the camera or farther back.
"""
from __future__ import annotations

import cv2
import numpy as np

# Points closer together than this fraction of the head size are collapsed:
# points that have converged carry no extra information, and a cluster
# would weight one part of the face more heavily than the rest.
PRUNING_CELL_FRACTION = 0.02

# A candidate is skipped only when an existing point is already within this
# fraction of the head size on BOTH axes -- i.e. genuinely nearby. Rejecting
# on either axis alone would discard symmetric features, which share a
# coordinate (the two nostrils sit at the same height, the midline points
# share an x). Established points already carry motion history, so they're
# preferred over a fresh candidate sitting on top of one.
MIN_ADD_DISTANCE_FRACTION = 0.03

# Average adult face height/width ratio: the cull region is an ellipse this
# much taller than it is wide, so it fits an actual face shape rather than
# a circle.
FACE_ASPECT_RATIO = 1.3

_LK_PARAMS = dict(
    winSize=(21, 21),
    maxLevel=2,
    criteria=(cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 30, 0.03),
    # Rejects low-texture tracks that tend to wander; well below OpenCV's
    # own default of 1e-4.
    minEigThreshold=0.001,
)


def _empty() -> np.ndarray:
    return np.empty((0, 2), dtype=np.float32)


def prune_points(points, prev_points, status, nose, head_size):
    """Drops points optical flow lost, collapses near-duplicates, and culls
    points that have drifted off the face -- in that order.

    Optical flow can report success for a point that has visibly diverged,
    so the region cull, not `status`, is the real backstop against a
    runaway point dragging the average off course. Returns the surviving
    (current, previous) arrays.
    """
    keep = np.asarray(status).reshape(-1) == 1
    points = np.asarray(points, dtype=np.float32).reshape(-1, 2)[keep]
    prev_points = np.asarray(prev_points, dtype=np.float32).reshape(-1, 2)[keep]

    cell = max(head_size * PRUNING_CELL_FRACTION, 1.0)
    seen_cells = {}
    for index, (x, y) in enumerate(points):
        seen_cells[(int(x // cell), int(y // cell))] = index
    if seen_cells:
        unique = sorted(seen_cells.values())
        points = points[unique]
        prev_points = prev_points[unique]

    if len(points):
        nose_x, nose_y = nose
        distances = np.hypot(
            (points[:, 0] - nose_x) * FACE_ASPECT_RATIO, points[:, 1] - nose_y
        )
        inside = distances <= head_size
        points = points[inside]
        prev_points = prev_points[inside]

    return points, prev_points


def should_add_point(candidate, points, head_size) -> bool:
    """Whether a candidate is far enough from every tracked point, on BOTH
    axes, to be worth adding."""
    points = np.asarray(points, dtype=np.float32).reshape(-1, 2)
    if not len(points):
        return True
    min_distance = max(head_size * MIN_ADD_DISTANCE_FRACTION, 1.0)
    close_x = np.abs(points[:, 0] - candidate[0]) <= min_distance
    close_y = np.abs(points[:, 1] - candidate[1]) <= min_distance
    return not bool(np.any(close_x & close_y))


def mean_movement(points, prev_points) -> tuple[float, float]:
    """Average frame-to-frame movement across every tracked point, in
    camera pixels."""
    points = np.asarray(points, dtype=np.float32).reshape(-1, 2)
    prev_points = np.asarray(prev_points, dtype=np.float32).reshape(-1, 2)
    if not len(points):
        return 0.0, 0.0
    delta = points - prev_points
    return float(delta[:, 0].mean()), float(delta[:, 1].mean())


class PointTracker:
    """Tracks a small set of face points across frames via optical flow."""

    def __init__(self) -> None:
        self._prev_gray = None
        self._points = _empty()
        self._prev_points = _empty()
        self._movement = (0.0, 0.0)

    @property
    def point_count(self) -> int:
        return len(self._points)

    def reset(self) -> None:
        """Drops every point and the previous frame, so the next update
        starts fresh. Called when the face is lost or a camera read fails --
        either way the next frame has no valid reference to compare against."""
        self._prev_gray = None
        self._points = _empty()
        self._prev_points = _empty()
        self._movement = (0.0, 0.0)

    def update(self, gray, nose, head_size, candidates) -> None:
        if self._prev_gray is not None and len(self._points):
            tracked, status, _err = cv2.calcOpticalFlowPyrLK(
                self._prev_gray, gray,
                self._points.reshape(-1, 1, 2), None,
                **_LK_PARAMS,
            )
            self._points, self._prev_points = prune_points(
                tracked.reshape(-1, 2), self._points, status, nose, head_size
            )
        else:
            self._prev_points = self._points.copy()

        # Measured before seeding: a point added this frame has no movement
        # to report, and averaging its zero in would understate the frame by
        # 1/N -- which shows up as intermittent cursor-speed dips.
        self._movement = mean_movement(self._points, self._prev_points)

        for candidate in candidates:
            if should_add_point(candidate, self._points, head_size):
                self._points = np.vstack(
                    [self._points, np.array([candidate], dtype=np.float32)]
                )
                self._prev_points = np.vstack(
                    [self._prev_points, np.array([candidate], dtype=np.float32)]
                )

        self._prev_gray = gray

    def get_movement(self) -> tuple[float, float]:
        """The movement measured by the last `update`, in camera pixels."""
        return self._movement

File: apps/desktop/src/renderer/tracking/pointTracker.ts
// Type-only: @techstark/opencv-js's default export is a Promise<CV> at
// runtime (the WASM module isn't ready synchronously on import), so this
// file must never touch it as a value at module scope -- that crashes on
// import (see task-15-report.md). PointTracker instead receives the already
// -resolved `cv` namespace via constructor injection from index.ts, which
// awaits it once. `typeof import(...)` is a pure type query (no runtime
// import emitted); a plain `import type CvNamespace from "..."` default
// import doesn't work here because this package has no `export default` in
// its .d.ts, so TS resolves the synthetic default to a *namespace*, and
// namespaces can't be used as ordinary type annotations (TS2709).
type CvNamespace = typeof import("@techstark/opencv-js");
import type { Mat as CvMat } from "@techstark/opencv-js";

export const PRUNING_CELL_FRACTION = 0.02;
export const MIN_ADD_DISTANCE_FRACTION = 0.03;
export const FACE_ASPECT_RATIO = 1.3;

export type Point = [number, number];

export function prunePoints(
  points: Point[],
  prevPoints: Point[],
  status: number[],
  nose: Point,
  headSize: number
): { points: Point[]; prevPoints: Point[] } {
  const keptIdx: number[] = [];
  status.forEach((s, i) => { if (s === 1) keptIdx.push(i); });
  let curr = keptIdx.map((i) => points[i]);
  let prev = keptIdx.map((i) => prevPoints[i]);

  const cell = Math.max(headSize * PRUNING_CELL_FRACTION, 1.0);
  const seenCells = new Map<string, number>();
  curr.forEach(([x, y], index) => {
    seenCells.set(`${Math.floor(x / cell)}:${Math.floor(y / cell)}`, index);
  });
  if (seenCells.size) {
    const unique = Array.from(seenCells.values()).sort((a, b) => a - b);
    curr = unique.map((i) => curr[i]);
    prev = unique.map((i) => prev[i]);
  }

  if (curr.length) {
    const [noseX, noseY] = nose;
    const keep: number[] = [];
    curr.forEach(([x, y], index) => {
      const distance = Math.hypot((x - noseX) * FACE_ASPECT_RATIO, y - noseY);
      if (distance <= headSize) keep.push(index);
    });
    curr = keep.map((i) => curr[i]);
    prev = keep.map((i) => prev[i]);
  }

  return { points: curr, prevPoints: prev };
}

export function shouldAddPoint(candidate: Point, points: Point[], headSize: number): boolean {
  if (!points.length) return true;
  const minDistance = Math.max(headSize * MIN_ADD_DISTANCE_FRACTION, 1.0);
  return !points.some(
    ([x, y]) => Math.abs(x - candidate[0]) <= minDistance && Math.abs(y - candidate[1]) <= minDistance
  );
}

export function meanMovement(points: Point[], prevPoints: Point[]): [number, number] {
  if (!points.length) return [0.0, 0.0];
  let sumX = 0, sumY = 0;
  for (let i = 0; i < points.length; i++) {
    sumX += points[i][0] - prevPoints[i][0];
    sumY += points[i][1] - prevPoints[i][1];
  }
  return [sumX / points.length, sumY / points.length];
}

// These two don't depend on cv at all, so (unlike LK_WIN_SIZE/LK_CRITERIA
// below) they stay as plain module-scope constants.
const LK_MAX_LEVEL = 2;
const LK_MIN_EIG_THRESHOLD = 0.001;

function toMat(cv: CvNamespace, points: Point[]): CvMat {
  const mat = new cv.Mat(points.length, 1, cv.CV_32FC2);
  for (let i = 0; i < points.length; i++) {
    mat.data32F[i * 2] = points[i][0];
    mat.data32F[i * 2 + 1] = points[i][1];
  }
  return mat;
}

function fromMat(mat: CvMat): Point[] {
  const out: Point[] = [];
  for (let i = 0; i < mat.rows; i++) {
    out.push([mat.data32F[i * 2], mat.data32F[i * 2 + 1]]);
  }
  return out;
}

export class PointTracker {
  private readonly winSize: InstanceType<CvNamespace["Size"]>;
  private readonly criteria: InstanceType<CvNamespace["TermCriteria"]>;
  private prevGray: CvMat | null = null;
  private points: Point[] = [];
  private prevPoints: Point[] = [];
  private movement: [number, number] = [0, 0];

  constructor(private readonly cv: CvNamespace) {
    this.winSize = new cv.Size(21, 21);
    this.criteria = new cv.TermCriteria(cv.TermCriteria_EPS + cv.TermCriteria_COUNT, 30, 0.03);
  }

  get pointCount(): number {
    return this.points.length;
  }

  reset(): void {
    this.prevGray?.delete();
    this.prevGray = null;
    this.points = [];
    this.prevPoints = [];
    this.movement = [0, 0];
  }

  update(gray: CvMat, nose: Point, headSize: number, candidates: Point[]): void {
    const cv = this.cv;
    if (this.prevGray && this.points.length) {
      const prevPtsMat = toMat(cv, this.points);
      const nextPtsMat = new cv.Mat();
      const status = new cv.Mat();
      const err = new cv.Mat();
      cv.calcOpticalFlowPyrLK(
        this.prevGray, gray, prevPtsMat, nextPtsMat, status, err,
        this.winSize, LK_MAX_LEVEL, this.criteria, 0, LK_MIN_EIG_THRESHOLD
      );

      const statusArr = Array.from(status.data);
      const { points, prevPoints } = prunePoints(fromMat(nextPtsMat), this.points, statusArr, nose, headSize);
      this.points = points;
      this.prevPoints = prevPoints;

      prevPtsMat.delete();
      nextPtsMat.delete();
      status.delete();
      err.delete();
    } else {
      this.prevPoints = [...this.points];
    }

    this.movement = meanMovement(this.points, this.prevPoints);

    for (const candidate of candidates) {
      if (shouldAddPoint(candidate, this.points, headSize)) {
        this.points = [...this.points, candidate];
        this.prevPoints = [...this.prevPoints, candidate];
      }
    }

    this.prevGray?.delete();
    this.prevGray = gray.clone();
  }

  getMovement(): [number, number] {
    return this.movement;
  }
}
